VSTSRestApiSamples/GettingStarted/Authentication.cs

```diff
@@ -1,55 +1,203 @@
+                client.DefaultRequestHeaders.Add("X-TFS-FedAuthRedirect", "Suppress");
+                client.DefaultRequestHeaders.Authorization = authHeader;
 
                 // connect to the REST endpoint            
                 HttpResponseMessage response = client.GetAsync("_apis/projects?stateFilter=All&api-version=2.2").Result;
-          
+
                 // check to see if we have a succesfull respond
                 if (response.IsSuccessStatusCode)
                 {
-                    // set the viewmodel from the content in the response
                     viewModel = response.Content.ReadAsAsync<ListofProjectsResponse.Projects>().Result;
                     // var value = response.Content.ReadAsStringAsync().Result;
                 }
+                else if (response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
+                {
+                    // This often occurs if the token has expired.
+                    // Acquire an updated token via the AuthenticationContext
+                    throw new UnauthorizedAccessException();
+                }
+                else
+                {
+                    Console.WriteLine("{0}:{1}", response.StatusCode, response.ReasonPhrase);
+                }
 
                 viewModel.HttpStatusCode = response.StatusCode;
 
                 return viewModel;
             }
+        }
 
-        }        
+        private static string GetAccountTenant(string vstsAccountName)
+        {
+            string tenant = null;
+            using (var client = new HttpClient())
+            {
+                client.BaseAddress = new Uri(String.Format("https://{0}.visualstudio.com", vstsAccountName));
+                client.DefaultRequestHeaders.Accept.Clear();
+                client.DefaultRequestHeaders.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));
+                client.DefaultRequestHeaders.Add("User-Agent", "VSTSAuthSample-AuthenticateADALNonInteractive");
+                client.DefaultRequestHeaders.Add("X-TFS-FedAuthRedirect", "Suppress");
+                HttpResponseMessage response = client.GetAsync("_apis/connectiondata").Result;
 
-    }
-}
+                try { 
+                    // Get the tenant from the Login URL
+                    var wwwAuthenticateHeaderResults = response.Headers.WwwAuthenticate.ToList();
+                    var bearerResult = wwwAuthenticateHeaderResults.Where(p => p.Scheme == "Bearer");
+                    foreach (var item in wwwAuthenticateHeaderResults)
+                    {
+                        if (item.Scheme.StartsWith("Bearer"))
+                        {
+                            tenant = item.Parameter.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries)[2];
+                            break;
+                        }
+                    }
+                }
+                catch (Exception ex)
```

Comment:
What exception are you catching here? Is this the case when there is no www-authenticate header, or it's not in the AAD format? If so, we should be writing a regex to parse this out without throwing exceptions. The return type of this method really should be a Guid, since tenant IDs are Guids. Guid.Empty represents MSAs.